Implementation of Gaussian, GaussianMixture, ParticleSet and related

xEnVrE opened the pull request:
Within this PR:
- Class `GaussianMixture` has now methods to access to mean, covariance and weight of the i-th component
- Class `Gaussian` re-implemented as inheriting class from class `GaussianMixture`
- Class `GaussianRef` removed
- Implemented class `ParticleSet` as inheriting class from class `GaussianMixture`
- Use class `ParticleSet` within classes `PFPrediction`, `PFPredictionDecorator`, `PFCorrection`, `PFCorrectionDecorator`, `DrawParticles`, `UpdateParticles`, `Resampling`, `ResamplingWithPrior`, `ParticleSetInitialization`, `InitSurveillanceAreaGrid` and `SIS`
- Method `ResamplingWithPrior::resample` heavily changed (due to use of ParticleSet)
- Method `SIS::SIS` take the state size as argument (required to initialize ParticleSet)
- Update `test_Gaussian`, `test_SigmaPointUtils `(since use class Gaussian)
- Update `test_SIS` (since use class `SIS`)
- Update `test_SIS_Decorators` (since use classes `PFPredictionDecorator`, `PFCorrectionDecorator` and `SIS`)
- Increase version to `0.8.101` since breaking API
- Update `CHANGELOG.md`

xEnVrE committed:
Implement class ParticleSet

xEnVrE committed:
Use ParticleSet within class PFPrediction

xEnVrE committed:
Use ParticleSet within class PFPredictionDecorator

xEnVrE committed:
Use ParticleSet within class PFCorrection

xEnVrE committed:
Use ParticleSet within class PFCorrectionDecorator

xEnVrE committed:
Use ParticleSet within class DrawParticles

xEnVrE committed:
Use ParticleSet within class UpdateParticles

xEnVrE committed:
Use ParticleSet within class Resampling

xEnVrE committed:
Use ParticleSet within class ParticleSetInitialization

xEnVrE committed:
Use ParticleSet within class InitSurveillanceAreaGrid

xEnVrE committed:
Change implementation of class SIS in order to use ParticleSet

xEnVrE committed:
Update test_SIS

xEnVrE committed:
Update test_SIS_Decorators

xEnVrE committed:
Use ParticleSet within class ResamplingWithPrior

xEnVrE committed:
Reimplement GaussianMixture with access to particle-wise data

xEnVrE committed:
Reimplement Gaussian as inheriting class from GaussianMixture

xEnVrE committed:
Remove class GaussianRef

xEnVrE committed:
Use new interface of Gaussian in test_Gaussian

xEnVrE committed:
Fix typo in test_Gaussian

xEnVrE committed:
Use new interface of Gaussian in sigma_point::unscented_transform

xEnVrE committed:
Update test_SigmaPointUtils

xEnVrE committed:
Update version to 0.8.101

xEnVrE committed:
Update CHANGELOG.md

claudiofantacci commented:
@xEnVrE meanwhile I'm reviewing this PR, can you please rebase onto devel to update Travis script?

xEnVrE commented:
@claudiofantacci rebased on last devel

claudiofantacci commented on src/BayesFilters/include/BayesFilters/Gaussian.h:
```diff
@@ -10,41 +12,29 @@ namespace bfl {
+       since a Gaussian is a 1-component GaussianMixture.
+       Hence it is better to return a Ref<VectorXd> as the mean,
+       rather than a Ref<MatrixXd>, and a double& as the weight,
+       rather than a Ref<VectorXd>&. */
```
For comments, please use the following format:
```cpp
/**
 * First line here
 * Next line here
 * Final line here 
 */
```

claudiofantacci commented on test/test_Gaussian/main.cpp:
```diff
@@ -71,70 +71,70 @@ int main()
-        if (!((gaussian_3d.mean(0) == 1) &&
-              (gaussian_3d.mean(1) == 2) &&
-              (gaussian_3d.mean(2) == 3)))
+        if (!((gaussian_3d.mean()(0) == 1) &&
```
~~I don't really like this syntax, but let's keep it like this for now.~~
~~I suggest to add element-wise accessor to `GaussianMixture` and `Gaussian`, for example.~~
```cpp
gaussian_3d(0) = 2;
```
~~implies the first element of the mean vector;~~
```cpp
gaussian_mixture_3d(3, 0) = 2;
```
~~implies the first element of the mean vector of the 3rd gaussian component of the mixture.~~
~~Same reasoning with `ParticleSet`/`Particle`.~~

~~I would also do the same for the covariances.~~

xEnVrE replied:
Implemented accessors for `Gaussian` and `GaussianMixture` (both mean and covariance).

As regards `ParticleSet` we should decide how to disambiguate between element-wise access to state and mean.

claudiofantacci replied:
I correct my previous examples that are wrong! I apologize.

I don't really like this syntax, but let's keep it like this for now.
I suggest to add element-wise accessor to `GaussianMixture` and `Gaussian`, for example.
```cpp
gaussian_3d.mean(0) = 2;
gaussian_3d.covariance(0, 0) = 2;
```
implies the first element of the mean vector and the `(0, 0)` element of the covariance;
```cpp
gaussian_mixture_3d.mean(3, 0) = 2;
gaussian_mixture_3d.covariance(3, 0, 0) = 2;
```
implies the first element of the mean vector and the `(0, 0)` element of the covariance of the 3rd gaussian component of the mixture.
Same reasoning with `ParticleSet`.

I would also do the same for the covariances.

xEnVrE replied:
Now it should be ok.

I also had to re-implement `Gaussian::covariance()` since the new method `Gaussian::covariance(const std:size_t i, const std::size_t j)` hides `GaussianMixture::covariance()` which was ok also for a Gaussian since the matrix of covariances shrinks to a single covariance matrix.

As regards `ParticleSet`, the methods for mean/covariance are inherited from `GaussianMixture` while I implemented element-wise access to the i-th state.

claudiofantacci commented on test/test_Gaussian/main.cpp:
```diff
@@ -24,7 +24,7 @@ int main()
         }
 
-        if (!catch_exception)
+        if (catch_exception)
```
Please remove the `catch_expetion` variable and move the `return EXIT_FAILURE` inside the `catch` block.

claudiofantacci commented on src/BayesFilters/include/BayesFilters/ParticleSet.h:
```diff
@@ -0,0 +1,36 @@
+
+    virtual ~ParticleSet() noexcept;
+
+    ParticleSet operator+(const ParticleSet& rhs);
```
I would suggest to also implement `operator+=` and to follow guidelines reported [here](https://en.cppreference.com/w/cpp/language/operators).

xEnVrE replied:
Done, however I decided not to put the symmetric operator

`ParticleSet operator+(ParticleSet lhs, const ParticleSet& rhs)`

as a member function because in that case it is required
- to put the definition of the function within the class body, i.e., within the header file
- to declare it as friend (but in this case friendship is not required)

claudiofantacci commented on src/BayesFilters/src/ResamplingWithPrior.cpp:
```diff
@@ -50,37 +50,50 @@ ResamplingWithPrior& ResamplingWithPrior::operator=(ResamplingWithPrior&& resamp
+    ParticleSet tmp_particles(num_resample_particles, cor_particles.dim_linear, cor_particles.dim_circular);
     int j = 0;
-    for (int i: sort_indices(cor_weights))
+    for (std::size_t i :sort_indices(cor_particles.weight().cast<float>()))
```
Missing space after `:`.

claudiofantacci reviewed (changes requested):
LGTM!

Thanks for the very good PR.
I have only minor comments and suggestions.
After addressing them, I'll merge the PR.

Should you need to add/remove/modify anything within this PR, remember to update the `CHANGELOG`.

claudiofantacci commented on src/BayesFilters/src/ResamplingWithPrior.cpp:
```diff
@@ -50,37 +50,50 @@ ResamplingWithPrior& ResamplingWithPrior::operator=(ResamplingWithPrior&& resamp
 
-void ResamplingWithPrior::resample(const Ref<const MatrixXf>& pred_particles, const Ref<const VectorXf>& cor_weights,
-                                   Ref<MatrixXf> res_particles, Ref<VectorXf> res_weights, Ref<VectorXf> res_parents)
+void ResamplingWithPrior::resample(const ParticleSet& cor_particles, ParticleSet& res_particles, Ref<VectorXf> res_parents)
```
Change `res_parents` type to `VectorXi`.

xEnVrE committed:
Change format of comments in Gaussian.h

xEnVrE committed:
Update test_Gaussian

Return within the catch block

xEnVrE committed:
Add missing space in ResamplingWithPrior::resample

xEnVrE committed:
Use VectorXi instead of VectorXf to represents particle parents in method Resampling::resample

xEnVrE committed:
Use VectorXi instead of VectorXf to represents particle parents in method ResamplingWithPrior::resample

xEnVrE committed:
Use VectorXi instead of VectorXf to represents particle parents in method method SIS::filteringStep

xEnVrE committed:
Implement compound ParticleSet::operator+= and implement non-member operator+(ParticleSet, const ParticleSet&)

xEnVrE commented:
Class `Particle` is not used at all and I left the original implementation as in the latest `devel`. Should we remove the files associated to it?

claudiofantacci commented:
> Class `Particle` is not used at all and I left the original implementation as in the latest `devel`. Should we remove the files associated to it?

Let's delete it!

xEnVrE committed:
Add methods to access to single elements of the mean and covariance of a Gaussian class

xEnVrE committed:
Update test_Gaussian

Use accessors to single elements of the mean/covariance of a Gaussian

xEnVrE committed:
Add methods to access to single elements of the mean and covariance of a component of a GaussianMixture class

xEnVrE committed:
Update test_Gaussian

Use accessors to single elements of the mean/covariance of a component of a GaussianMixture

xEnVrE committed:
Add methods to access to single elements of the state of a ParticleSet class

xEnVrE committed:
Remove class Particle

xEnVrE committed:
Update CHANGELOG.md

claudiofantacci commented on test/test_Gaussian/main.cpp:
```diff
@@ -22,10 +22,9 @@ int main()
         {
             catch_exception = true;
```
Almost there!
Please delete this line and `catch_exception` declaration some lines above as we are no longer using it.

After this, I merge the PR 🎉

xEnVrE replied:
Sorry, forgot to remove them!

xEnVrE committed:
Remove unused lines of code in test_Gaussian/main.cpp

claudiofantacci commented:
LGTM!

Thanks for the big effort and contribution @xEnVrE!
